### ISSUE TITLE
Collect: easier access to SMS-fallback workflow

### ISSUE THREAD

**Issue by lognaturel** [getodk/roadmap#24]
### Background and strategic fit
Collect 1.16 adds SMS submissions but the process is too manual to fit the needs of users who’ve tried it. Typically, users want submissions to fall back to SMS without user intervention if Internet is not available.

### Goals
- Improve workflow for users who primarily use SMS as a backup protocol (because that’s what we have feedback about)
- Do not hamstring users who use SMS as a primary transport.
- Simplify the user experience around SMS to capture the 90% use-case
- Find the narrowest path to get this feature working (because we don’t have a ton of user feedback yet)

### Non-goals
- Solving the issue of re-sending SMS submissions because it’s hard to know when to retry, how often to retry, etc.

### User stories
- As an supervisor, when my enumerators are in an area with no Internet, but with SMS, I would like the key indicators in the form to be sent so I can have that data immediately.
- As an enumerator, I would like this data sent on form finalize without me directly selecting the file for submission.
- As an enumerator, I would like the complete filled form to be sent when Internet is back without manual intervention.
- As an enumerator, I would like to know if SMS submission fails because the data is time-sensitive and I can potentially improve my chances of a successful submission just by moving a bit.

### Current functionality
From Server settings, a user can set what the manual send button on Send Finalized Form does and set an SMS destination phone number. 

From Form management settings, a user can auto-send to an HTTP server. There is no way to trigger an auto-send via SMS.

<img width="200" alt="screen shot 2018-09-07 at 3 20 16 pm" src="https://user-images.githubusercontent.com/967540/45245571-cb072a00-b2b1-11e8-915d-829e296093dd.png"><img width="200" alt="screen shot 2018-09-07 at 3 20 35 pm" src="https://user-images.githubusercontent.com/967540/45245593-dbb7a000-b2b1-11e8-86fb-e0e3a039ef81.png"><img width="200" alt="screen shot 2018-09-07 at 3 21 03 pm" src="https://user-images.githubusercontent.com/967540/45245611-eeca7000-b2b1-11e8-9c95-fe5bfa0ad683.png">

### User interaction and design
As a supervisor or enumerator, I need to be able to identify that I want SMS submission attempted after an Internet submission fails.

Conceptually, the desired flow on send is:
![flow](https://user-images.githubusercontent.com/967540/45068404-23ed7d00-b07c-11e8-88c6-3e327d0d174d.png)

Sending an SMS should only be attempted if the form has SMS tags and an SMS hasn't been sent before. That is, only one SMS should ever be sent by this workflow. The user can still trigger a resend manually if the send button sends over SMS.

#### Settings changes
- Move “Send submissions via” currently in server settings to “Form management” > “Form submission” right before auto send and rename it to “Manual send button sends over” (options Wifi/Cellular (recommended), SMS, Select on Send). This makes it clear that this setting changes the meaning of the manual button.
- Leave “SMS submission settings” for setting destination phone number in Server settings but always enable it (currently it’s only enabled if “Send submissions via” is set to SMS or Select on Send)
- Rename “Auto send” to “Wifi/Cellular auto send” to make it clear there are no auto send SMS options there.
- Add toggle below “Wifi/Cellular auto send” that says “Fall back to SMS” with description text “Try SMS after Wifi/Cellular send failure”. This toggle is only available if there is an SMS destination phone number set in “Server”. If this is on, every time a send attempt is made (triggered manually, or on auto-send), if that attempt fails, an SMS is attempted. It can be enabled even if auto send is off.

![settings](https://user-images.githubusercontent.com/967540/45068490-834b8d00-b07c-11e8-92b1-5a7a28f272e2.png)

**Typical settings combinations**

*Default*
Nothing happens automatically. To send a form, the user taps the Send button on the Send Finalized Form screen and the submission(s) are sent over HTTP.
```
Manual send button sends over: Wifi/Cellular (recommended)
Wifi/Cellular auto send: Off
Fall back to SMS: Off
```

*HTTP autosend*
When a user finishes filling a form, that form sends immediately over HTTP if an Internet connection is available. Otherwise, it is enqueued and sends when a connection becomes available. The user can also go to Send Finalized Form and try sending form(s) manually.
```
Manual send button sends over: Wifi/Cellular (recommended)
Wifi/Cellular auto send: Wifi, cellular or both
Fall back to SMS: Off
```

*SMS-only*
Nothing happens automatically. To send a form, the user taps the Send button on the Send Finalized Form screen and the submission(s) are sent over SMS.
```
Manual send button sends over: SMS
Wifi/Cellular auto send: Off
Fall back to SMS: Off
Phone number must be set in Server settings
```

*Auto send with SMS Fallback*
When a user finishes filling a form, that form sends immediately over HTTP if an Internet connection is available. If the connection is not available or the connection fails, and the form contains SMS tags and an SMS hasn't been sent, an SMS is sent. Then the form is enqueued for HTTP submission and sends when a connection becomes available.
```
Manual send button sends over: any
Wifi/Cellular auto send: Wifi/Cellular or Both
Fall back to SMS: on
Phone number must be set in Server settings
```

#### Form sent status change
Currently, forms that have been sent via SMS or via Internet have the same `SUBMITTED` status in the database. Additionally, auto-send does not do a true enqueue, it simply sends anything that doesn’t have a `SUBMITTED` status. This means there’s no way to sequence an Internet submission after an SMS submission -- the filled form looks like it’s been sent already. To support this sequencing, introduce an `SMS_SUBMITTED` status. Filled forms that have the `SMS_SUBMITTED` status will still be queued for auto send over Internet.

#### Send Finalized Form Interface change
By default, the Send Finalized Form screen only shows unsent forms. Once a form is sent, it disappears from that screen (but can be shown using the sent and unsent view option from the overflow menu). With the introduction of the `SMS_SUBMITTED` status and workflows that may involve submitting first over SMS and then over Internet we need to rethink the most appropriate forms to show in the Send Finalized Form interface.

If the manual send button is configured to send over SMS and autosend is off, hide `SMS_SUBMITTED` forms because they are considered fully submitted. In all other cases, show `SMS_SUBMITTED` forms on the unsent screen because they still need to be sent over Internet.

### Additional implementation details
- a database column will need to be added to keep track of whether a submission has been successfully sent over SMS. This is necessary to send the SMS only once.

#### Add SMS auto-send option to existing auto-send menu rather than as separate toggle.
What this specification proposes is not a true auto-send because there is no retry so it seems potentially misleading to include there. It also seems this would result in 6 options: Wifi only, Cellular only, Wifi or Cellular and those same ones with SMS retry options.

**Comment by poVoq** [comment]
It would be also useful to only send an SMS if a specific answer is selected.

The user story would be:
General vulnerability assessment, but there might be some urgent to response to cases that should not wait for the general data analysis a while later. So if a certain question is answered with a specific response, send out a SMS as an alert with only the most relevant data.

Of course the full survey response should still be added to the database through the normal internet upload.

Edit: In fact this could also be a general mechanism with a separate, shorter form being populated with such kind of alerts.

**Comment by lognaturel** [comment]
Thanks for sharing this requirement, @poVoq, it makes a lot of sense.

I see two good options for achieving what you want with existing functionality: 1) only make questions with SMS tags relevant if a certain condition is true 2) show an SMS send prompt for the enumerator.

I think the first option is your best bet. The idea is to have all your normal questions and then have calculates that are only relevant under a specific condition. I haven't tried this form but it would look something like:

|   type    |       name        | compact_tag |              label               |   calculation    |        relevant        |
|-----------|-------------------|-------------|----------------------------------|------------------|------------------------|
| text      | phone_number      |             | Patient phone number             |                  |                        |
| decimal   | temperature_f     |             | Patient temperature in Fahrenheit |                  |                        |
|           |                   |             |                                  |                  |                        |
| calculate | sms_phone_number  | n           |                                  | ${phone_number}  | ${temperature_f} > 105 |
| calculate | sms_temperature_f | t           |                                  | ${temperature_f} | ${temperature_f} > 105 |

In that case, if temperature is below 105F, an SMS submission would not be generated. With the feature described in this specification, that means there would be no attempt to auto-send an SMS. On the other hand, if the recorded temperature was greater than 105F, an SMS would be generated and auto-sent if the settings are set to fall-back to SMS. I see that the case where an SMS is not generated is not explicitly called out by the spec and will add that.

Something else you might want to do before this auto-send is implemented is to include a question in the form that sends an SMS (see [the documentation](https://docs.opendatakit.org/launch-apps-from-collect/?highlight=uri_data#launch-apps-uri-data) and [a sample form](https://forum.opendatakit.org/t/sending-a-text-message-sms-or-email-from-a-form-in-collect/13874)). I would [conditionally show the question](https://docs.opendatakit.org/form-logic/#conditionally-showing-questions) if the specific indicator were set (same relevant idea as above).

The downside of this approach is that it does require the enumerator to manually tap "send" on the SMS. See [this forum post](https://forum.opendatakit.org/t/sending-a-text-message-sms-or-email-from-a-form-in-collect/13874/10) for more information on why it's currently set up this way. What we might be able to do is get information back from the SMS app to know whether the SMS was successfully sent and only let the user advance in the form if it was. If this sounds interesting to you, please post on that thread so it can be considered. The downside is that if an enumerator truly is in an area with no cell reception, they won't be able to make progress.

As a side note, I see from https://github.com/JhulFramework/PHPServer4ODK/issues/3 that you are interested in self-hosting a server. You may find [ODK Central](https://forum.opendatakit.org/t/odk-central-v0-2-beta/15005) an interesting option. Please continue the conversation on the forum if interested.

**Comment by lognaturel** [comment]
@shobhitagarwal1612 I added some text below the "flow" diagram specifying that an SMS should only be sent once and I noted that this will require adding a DB column to keep track of the state. Thank you. I also added a section at the very end about alternatives considered. If I misunderstood your suggestion, please let me know!

**Comment by alxndrsn** [comment]
Generally :+1: :+1: from me.  Some specific points addressed below:

> Leave “SMS submission settings” for setting destination phone number in Server settings but always enable it (currently it’s only enabled if “Send submissions via” is set to SMS or Select on Send)

I wonder if there is some way to make it clear _when_ this option is relevant if a user sets it.  Or perhaps flag it as irrelevant (e.g. via a tooltip) if no SMS submission option is currently selected.  But likely not important in a first iteration of these tweaks.


> Add toggle below “Wifi/Cellular auto send” that says “Fall back to SMS” with description text “Try SMS after Wifi/Cellular send failure”. This toggle is only available if there is an SMS destination phone number set in “Server”. If this is on, every time a send attempt is made (triggered manually, or on auto-send), if that attempt fails, an SMS is attempted. It can be enabled even if auto send is off.

If I'm reading correctly, there will be no "auto-send as SMS" option, but if this were desired, a user would need to enable "auto-send over wifi/cellular" and wait for that to fail every time.  Seems weird not to also have a "auto-send as SMS" directly available.


> To support this sequencing, introduce an `SMS_SUBMITTED` status.

It might be clearer to rename `SUBMITTED` to e.g. `SUBMITTED_HTTP` (at least in code, if not in the database) to avoid confusion.


> a database column will need to be added to keep track of whether a submission has been successfully sent over SMS. This is necessary to send the SMS only once.

If there's already a database field which notes whether a form has been `SUBMITTED` then assuming this is not a boolean field then I would guess this can also track the `SUBMITTED_SMS` status.


> What this specification proposes is not a true auto-send because there is no retry so it seems potentially misleading to include there. It also seems this would result in 6 options: Wifi only, Cellular only, Wifi or Cellular and those same ones with SMS retry options.

Or three options:

Auto send:

[ ] wifi
[ ] cellular
[ ] SMS

_(options will be tried in that order)_

-----

<img src="https://user-images.githubusercontent.com/967540/45068404-23ed7d00-b07c-11e8-88c6-3e327d0d174d.png"/>

> Sending an SMS should only be attempted if the form has SMS tags and an SMS hasn't been sent before. That is, only one SMS should ever be sent by this workflow.

By my reading, the text and diagram disagree on whether multiple SMS can be sent.

**Comment by admbtlr** [comment]
@alxndrsn Based on @lognaturel's comment above, I believe that the text supersedes the diagram, and that only one SMS will ever be sent

**Comment by lognaturel** [comment]
Thank you @alxndrsn for the comment and @adamvert for the extra ping!

> Seems weird not to also have a "auto-send as SMS" directly available.

I do agree with you and what you say is similar to @shobhitagarwal1612's complaint. To go in a bit more depth, the reason for no retry attempt is that (as far as I can tell) there isn't any great trigger for retrying an SMS failure. On the XML submission side, it makes sense to retry when there's a change on network availability. On the SMS side, I think the only option would be a time-based retry. We don't currently have a good way of communicating that kind of status so I think it could become quite confusing. But perhaps you have some alternative ideas? **EDIT: this is not quite accurate -- there would currently be retries but only as a response to an internet send failure**

> Or three options:
> Auto send:
> [ ] wifi
> [ ] cellular
> [ ] SMS

True that this is not bad. Maybe another way to communicate that SMS retry is different would be to make the third option something like `SMS (only tries once)`? **EDIT: This is not quite accurate -- there would currently be retries but only as a response to an internet send failure. Maybe `SMS (not guaranteed to send)`? The big difference is that once a network submission is made, we wouldn't send the SMS because for the projects we've spoken to, it wouldn't be needed if the network submission goes through.**

> It might be clearer to rename SUBMITTED to e.g. SUBMITTED_HTTP (at least in code, if not in the database) to avoid confusion.

👍 

> If there's already a database field which notes whether a form has been SUBMITTED then assuming this is not a boolean field then I would guess this can also track the SUBMITTED_SMS status.

The new database column would be to ensure an SMS is only auto-sent once. There's currently just a single status column and so let's say the auto-send via internet fails once, the status column would represent failure. Then an SMS submission would be attempted, if it succeeds, the status column represents SMS success. Then internet submission is attempted again, it fails again, the status column represents failure. Now an SMS submission would be attempted again. So I believe we need a separate status to know whether an SMS was ever successfully auto-sent and if so, to prevent an other SMS auto-send attempt.

I think it would be better to keep track of all send attempts in a separate database table so there's history available but that is a larger change that I'm not sure we want to prioritize now. Adding an SMS-specific column should address the problem without adding much complexity (and not closing the door to the history log I just described).

> By my reading, the text and diagram disagree on whether multiple SMS can be sent.
> I believe that the text supersedes the diagram, and that only one SMS will ever be sent

Yes, my apologies, I didn't update the beautiful diagram!  😉

**Comment by lognaturel** [comment]
One more thing that I now remember about why I don't feel great about SMS auto-send being a standalone option -- it's if ONLY SMS is selected that the lack of retry seems really problematic. I think it makes sense when it's in combination with a network send so we know the form will be sent somewhere but it doesn't feel right as its own auto-send option when it does in fact only try once. But perhaps the `SMS (only tries once)` wording described above would be a way to communicate that.

**Comment by lognaturel** [comment]
Discussed again at 10/17 TSC meeting and we concluded that:
- There is really no good way to trigger an automatic SMS retry because 
  - delivery reports are unreliable
  - there are no "cell network availability/strength" events we can trigger background jobs from
  - "cell network availability/strength" is not always a good indicator of whether a message can be sent
- The "fallback SMS" checkbox is a reasonable compromise for now
- We can specify that a fallback submission is only attempted once

The specification was approved.

**Comment by yanokwa** [comment]
@lognaturel Given that Google no longer allows SMS, I think we should close this issue.

**Comment by lognaturel** [comment]
Good thing we didn't build it! 😅😭

**Comment by MartijnR** [comment]
Does that mean that we don't actually have tools ourselves that use the compact submission representation (nor have plans to use this)?

**Comment by lognaturel** [comment]
Sigh, I suppose that is now the case, @MartijnR. Modulo some folks who really need SMS integration and are sideloading older Collect APKs or reviving the SMS code from a fork since it's still optimistically in there.

There have been requests to enable [sending compact records to servers](https://forum.opendatakit.org/t/submit-forms-to-aggregate-using-compact-tag-syntax/14669). This would require more spec work to define how exactly that blob would get sent.

**Comment by MartijnR** [comment]
Ok, thx. I believe Medic Mobile is also using it (is that right @garethbowen?). We could leave it in the spec for now imo.

**Comment by garethbowen** [comment]
Yes we're using SMS, but we use SMS only so we don't use a fallback process. And yes, also having to sideload it now to get around Google :sob:
